### PR TITLE
Corosync cmapctl gatherer

### PR DESCRIPTION
**Pull request by rtorrero**
This PR adds a gatherer to collect facts from the `corosync-cmapctl` binary and makes it available through our factsengine

Note that I had to move the utils to a separate package as I was incurring in cyclic dependencies when attempting it to use it in the internal package

### CODE REVIEW

File: internal/factsengine/gatherers/corosynccmapctl.go [hunk 1]
```diff
@@ -0,0 +1,45 @@
+package gatherers
+
+import (
+	log "github.com/sirupsen/logrus"
+	"github.com/trento-project/agent/internal/utils"
+)
+
+const (
+	CorosyncCmapCtlFactKey = "corosync-cmapctl"
```
Comment (arbulu89) [inline internal/factsengine/gatherers/corosynccmapctl.go:9]: In the package version gatherer I have used `_` as "splitter". We should agree on this hehe
`-` or `_`?

File: internal/factsengine/gatherers/corosynccmapctl.go [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
+	return &CorosyncCmapctlGatherer{
+		executor: Executor{},
+	}
+}
+
+func (s *CorosyncCmapctlGatherer) Gather(factsRequests []FactRequest) ([]Fact, error) {
+	facts := []Fact{}
+	log.Infof("Starting corosync-cmapctl facts gathering process")
+
+	corosyncCmapctl, err := s.executor.Exec( // nolint:gosec
```
Comment (arbulu89) [inline internal/factsengine/gatherers/corosynccmapctl.go:26]: Why are we disabling linting here? I think we can remove it safely as the arguments are constants

File: internal/factsengine/gatherers/corosynccmapctl.go [hunk 3]
```diff
@@ -0,0 +1,45 @@
...
+		"corosync-cmapctl", "-b")
+	if err != nil {
+		return facts, err
+	}
+
+	corosyncCmapctlMap := utils.FindMatches(`(?m)^(\S*)\s\(\S*\)\s=\s(.*)$`, corosyncCmapctl)
+
+	for _, factReq := range factsRequests {
+		if value, ok := corosyncCmapctlMap[factReq.Argument]; ok {
+			facts = append(facts, Fact{
```
Comment (arbulu89) [inline internal/factsengine/gatherers/corosynccmapctl.go:36]: Let's use the new `NewFactWithRequest` function, so we create the facts with all the needed fields

File: internal/factsengine/gatherers/corosynccmapctl.go [hunk 4]
```diff
@@ -0,0 +1,45 @@
...
+	corosyncCmapctl, err := s.executor.Exec( // nolint:gosec
+		"corosync-cmapctl", "-b")
+	if err != nil {
+		return facts, err
+	}
+
+	corosyncCmapctlMap := utils.FindMatches(`(?m)^(\S*)\s\(\S*\)\s=\s(.*)$`, corosyncCmapctl)
+
+	for _, factReq := range factsRequests {
+		if value, ok := corosyncCmapctlMap[factReq.Argument]; ok {
```
Comment (arbulu89) [inline internal/factsengine/gatherers/corosynccmapctl.go:35]: What if the value is not found?
We should handle this scenario. Maybe by now, we can put a log, but at some point we need to send back a fact with this data (maybe a new define hack).

File: internal/factsengine/gatherers/corosynccmapctl_test.go [hunk 5]
```diff
@@ -0,0 +1,112 @@
...
+)
+
+type CorosyncCmapctlTestSuite struct {
+	suite.Suite
+}
+
+func TestCorosyncCmapctlTestSuite(t *testing.T) {
+	suite.Run(t, new(CorosyncCmapctlTestSuite))
+}
+
```
Comment (arbulu89) [inline internal/factsengine/gatherers/corosynccmapctl_test.go:20]: Maybe we will need to have a test for the scenario that the wanted fact is not present (new unhandled scenario I comment in a previous comment)

**Review by arbulu89 — CHANGES_REQUESTED** [review]
Nice @rtorrero ,
You will need to rebase from main to get the latest changes, which will break some tests I'm afraid hehe

Besides that, once this or the `package_version` gatherer PRs are merge, we will need to rebase back again, as they have duplicated code for the command executor.

Finally, this new gatherer must be enabled in the `factsengine` code, as corosyncconf is right now. Otherwise we cannot really use it.

I would try to compile the binary and run to see that it actually works:
```
./trento-agent facts gather --gatherer corosync-cmapctl --argument runtime.config.totem.consensus
```

File: internal/factsengine/gatherers/corosynccmapctl.go (hunk 1 above)
Comment (rtorrero) [reply internal/factsengine/gatherers/corosynccmapctl.go:9]: it's not really splitting anything, but reflecting the name of the binary as-is in the same way `corosync.conf` is the fact key for the corosync configuration gatherer. I still understand the confusion and don't have a strong opinion on this, wdyt?

**Comment by rtorrero** [comment]
Thanks for the review, indeed I was missing enabling the gatherer in the facts engine. Seems to work correctly now:

```
INFO[2022-08-04 11:24:01] Using config file: /etc/trento/agent.yaml    
INFO[2022-08-04 11:24:01] Starting corosync-cmapctl facts gathering process 
INFO[2022-08-04 11:24:01] Requested corosync-cmapctl facts gathered    
INFO[2022-08-04 11:24:01] Gathered fact for "corosync-cmapctl" with argument "runtime.config.totem.consensus": 
INFO[2022-08-04 11:24:01] {
  "name": "runtime.config.totem.consensus",
  "value": "36000",
  "check_id": ""
} 
```

File: internal/factsengine/gatherers/corosynccmapctl.go [hunk 6]
```diff
@@ -0,0 +1,47 @@
...
+			fact := NewFactWithRequest(factReq, fmt.Sprint(value))
+			facts = append(facts, fact)
+		} else {
+			log.Warnf("%s gatherer: requested fact %s not found", CorosyncCmapCtlFactKey, factReq.Argument)
```
Comment (arbulu89) [inline internal/factsengine/gatherers/corosynccmapctl.go:41]: I have used `log.Errorf` in other gatherers, but I guess it doesn't really matter that much.
Once we know how to populate the failed fact gatherings we will need to come back and finish the job hehe

**Review by arbulu89 — APPROVED** [review]
Perfect!
Green light from my side

**Pull request merged**